openedx/core/djangoapps/programs/utils.py

```diff
@@ -126,6 +126,43 @@ def engaged_programs(self):
 
         return programs
 
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of strings, each representing a certificate course run key
+        """
+        course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
+        return [
+            unicode(certificate['course_key']) for certificate in course_run_certificates
+            if not certificate_api.is_passing_status(certificate['status'])
+        ]
+
+    def does_course_still_qualify_as_in_progress(self, now, enrolled_run_modes, course):
```

Comment:
I was thinking that _is_course_in_progress is checking if a course is in progress, which is different from whether a course is still counting towards a program. however, this method is in the program progress meter, so i suppose it is not used for courses individually. I could put the code there and maybe rename it to something like
_is_course_in_progress_and_still_counts_toward_program? 
it would easily be confusing if a course is in fact in progress but returned false because it is not upgradeable anymore for instance